File: lab9/lab9_1.m
clear all;
close all;
clc
% Parametry sygnału
fs = 8000;  % Częstotliwość próbkowania [Hz]
t = 1;      % Czas trwania [s]

% Częstotliwości i amplitudy harmonicznych
f1 = 34.2;  % Częstotliwość pierwszej harmonicznej [Hz]
A1 = -0.5;  % Amplituda pierwszej harmonicznej
f2 = 115.5; % Częstotliwość drugiej harmonicznej [Hz]
A2 = 1;     % Amplituda drugiej harmonicznej

% Tworzenie osi czasu
time = 0:1/fs:(t-1/fs);

% Generowanie sygnałów harmonicznych
harmonic1 = A1 * sin(2 * pi * f1 * time);
harmonic2 = A2 * sin(2 * pi * f2 * time);
SNR = [10, 20, 40];
dref = harmonic1 + harmonic2;
for k = 1:3
    d = awgn( dref, SNR(k), 'measured'); % WE: sygnał odniesienia dla sygnału x
    x = [ d(1) d(1:end-1) ]; % WE: sygnał filtrowany, teraz opóźniony d
    M = 100; % długość filtru
    mi = 0.00025;% współczynnik szybkości adaptacji

    %% Zwiększenie M - większy szum
    %% Zmniejszenie M - przesunięcie o fazę

    %% Zwiększenie mi - większy szum
    %% Zmniejszenie mi - wypłaszczanie sygnału

    y = []; e = []; % sygnały wyjściowe z filtra
    bx = zeros(M,1); % bufor na próbki wejściowe x
    h = zeros(M,1); % początkowe (puste) wagi filtru

    for n = 1 : length(x)
        bx = [ x(n); bx(1:M-1) ]; % pobierz nową próbkę x[n] do bufora
        y(n) = h' * bx; % oblicz y[n] = sum( x .* bx) – filtr FIR
        e(n) = d(n) - y(n); % oblicz e[n]
        h = h + mi * e(n) * bx; % LMS
        %h = h + mi * e(n) * bx /(bx'*bx); % NLMS
    end
    %% liczenie SNR
    N = length(dref);
    Ps = sum(dref.^2) / N;
    Pn = sum((dref-y).^2) / N;
    output_snr = 10 * log10(Ps / Pn);

    subplot(1, 3, k)
    plot( time, d, "r",time, y, "g", time, dref, "b--")
    title("SNR = "+ output_snr + "dB")
    legend( "Zaszumiany sygnał","Odszumiony sygnał", "Referencyjny sygnał")
    xlim([0.54, 0.55])
    ylim([0.5, 2])
end

File: lab9/lab9_1.py
import numpy as np
import matplotlib.pyplot as plt
# Parametry sygnału
fs = 8000  # Częstotliwość próbkowania [Hz]
t = 1  # Czas trwania [s]

# Częstotliwości i amplitudy harmonicznych
f1 = 34.2  # Częstotliwość pierwszej harmonicznej [Hz]
A1 = -0.5  # Amplituda pierwszej harmonicznej
f2 = 115.5  # Częstotliwość drugiej harmonicznej [Hz]
A2 = 1  # Amplituda drugiej harmonicznej

# Tworzenie osi czasu
timestep = 1 / fs  # Krok czasowy
time = np.arange(0, t, timestep)  # Oś czasu

# Generowanie sygnałów harmonicznych
harmonic1 = A1 * np.sin(2 * np.pi * f1 * time)
harmonic2 = A2 * np.sin(2 * np.pi * f2 * time)
dref = harmonic1 + harmonic2                                              # sygnał ”czysty” do porównania

target_snr_db = 20
t = np.linspace(1, 100, 1000)
x_volts = 10*np.sin(t/(2*np.pi))
x_watts = x_volts ** 2
sig_avg_watts = np.mean(x_watts)
sig_avg_db = 10 * np.log10(sig_avg_watts)
noise_avg_db = sig_avg_db - target_snr_db
noise_avg_watts = 10 ** (noise_avg_db / 10)

d = np.random.normal(dref, np.sqrt(noise_avg_watts)) # WE: sygnał  odniesienia dla sygnału x
x = np.append(d[0], d[:-1])                             #  WE: sygnał filtrowany, teraz opóźniony d
M = ???                                                 # długość filtru
mi = ???                                                #  współczynnik szybkości adaptacji

# sygnały wyjściowe z filtra
y = []
e = []
bx = np.zeros(M,1)                                      # bufor na próbki wejściowe x
h = np.zeros(M,1)                                       # początkowe (puste) wagi filtru

for n in range(1,x.length):
    bx[0] = np.insert(bx[:-1], [0], x[n])               # pobierz nową próbkę x[n] do bufora
    y[n] = h.transpose() * bx                           # oblicz y[n] = sum( x .* bx) – filtr FIR
    e[n] = d[n] - y[n]                                  # oblicz e[n]
    h = h + mi * e[n] * bx                              # LMS
    # h = h + mi * e[n] * bx /(bx.transpose() * bx)     # NLMS

print(len(time))

plt.plot(time, d)
plt.show()